crates/data-generator/src/blocks/output.rs

```diff
@@ -0,0 +1,586 @@
+                    .iter()
+                    .for_each(|prop| prop.to_tokens(&mut props));
+
+                // "LargeAmethystBud { facing: Facing::North, waterlogged: Waterlogged::True, }"
+                tokens.append(Group::new(Delimiter::Brace, props))
+            }
+        }
+    }
+
+    #[derive(Debug)]
+    pub struct BlockStateProperty<'a> {
+        pub field_name: &'a Ident, // instrument
+        pub enum_name: &'a Ident,  // NoteBlock
+        pub var_name: Ident,       // IronXylophone
+    }
+
+    impl ToTokens for BlockStateProperty<'_> {
+        fn to_tokens(&self, tokens: &mut TokenStream) {
+            let field_name = &self.field_name;
+            let enum_name = &self.enum_name;
+            let var_name = &self.var_name;
+
+            // "instrument: NoteBlock::IronXylophone,"
+            tokens.extend(quote!(#field_name: #enum_name::#var_name,))
+        }
+    }
+
+    impl ToTokens for Id {
+        fn to_tokens(&self, tokens: &mut TokenStream) {
+            tokens.append(Literal::u32_unsuffixed(self.0))
+        }
+    }
+}
+
+mod prop_enum {
+    use proc_macro2::{Delimiter, Group, Ident, Literal, TokenStream};
+    use quote::{quote, ToTokens, TokenStreamExt as _};
+
+    use crate::{blocks::model::PropertyVariants, name::Name};
+
+    pub struct PropertyEnum<'a> {
+        pub prop_name: &'a Ident, // Powered
+        pub prop_vars: Vec<(Ident, Option<Literal>)>,
+    }
+
+    impl<'a> PropertyEnum<'a> {
+        pub fn new(name: &'a Name, variants: &PropertyVariants) -> Self {
+            Self {
+                prop_name: name.pascal_case(),
+                prop_vars: variants
+                    .as_enum()
+                    .into_iter()
+                    .map(|(name, ord)| {
+                        (
+                            name.into_pascal_case(),
+                            ord.map(|x| Literal::u8_unsuffixed(x)),
+                        )
+                    })
+                    .collect(),
+            }
+        }
+    }
+
+    impl ToTokens for PropertyEnum<'_> {
+        fn to_tokens(&self, tokens: &mut TokenStream) {
+            let derive = if self
+                .prop_vars
+                .first()
+                .map(|(_, ord)| ord.is_some())
+                .unwrap_or(false)
+            {
+                quote!(Ord, PartialOrd,)
+            } else {
+                TokenStream::new()
+            };
+
+            let prop_name = &self.prop_name;
+            // "pub enum BubbleCoralFan"
+            tokens.extend(quote! {
+                #[derive(Debug, #derive Eq, PartialEq, Copy, Clone, Hash)]
+                pub enum #prop_name
+            });
+
+            let mut stream = TokenStream::new();
+            self.prop_vars.iter().for_each(|(ident, ord)| {
+                match ord {
+                    Some(ord) => quote!(#ident = #ord,),
+                    None => quote!(#ident,),
+                }
+                .to_tokens(&mut stream)
```

Comment:
`quote!()` repetition can help you out here

Reply:
Done; see [`d51bd73:/crates/data-generator/src/blocks/output.rs`](https://github.com/rjacraft/rjacraft/blob/d51bd73625394553e9b6bc0b615eb56ce26835b7/crates/data-generator/src/blocks/output.rs#L383-L406):
https://github.com/rjacraft/rjacraft/blob/d51bd73625394553e9b6bc0b615eb56ce26835b7/crates/data-generator/src/blocks/output.rs#L383-L406